frontend/src/core/codemirror/language/markdown.ts

```diff
@@ -0,0 +1,169 @@
+/* Copyright 2023 Marimo. All rights reserved. */
+import { Extension } from "@codemirror/state";
+import { LanguageAdapter } from "./types";
+import { markdown } from "@codemirror/lang-markdown";
+import { languages } from "@codemirror/language-data";
+import { parseMixed } from "@lezer/common";
+import { python, pythonLanguage } from "@codemirror/lang-python";
+import dedent from "dedent";
+import { logNever } from "@/utils/assertNever";
+
+const prefixKinds = ["", "f", "r", "fr", "rf"] as const;
+type PrefixKind = (typeof prefixKinds)[number];
+
+const quoteKinds = [
+  ['"""', '"""'],
+  ["'''", "'''"],
+  ['"', '"'],
+  ["'", "'"],
+];
+// explode into all combinations
+const pairs = prefixKinds.flatMap((prefix) =>
+  quoteKinds.map(([start, end]) => [prefix + start, end])
+);
+
+const regexes = pairs.map(
+  ([start, end]) =>
+    // mo.md( + any number of spaces + start + capture + any number of spaces + end)
+    [
+      start,
+      new RegExp(`^mo\\.md\\(\\s*${start}(.*)${end}\\s*\\)$`, "s"),
+    ] as const
+);
+
+/**
+ * Language adapter for Markdown.
+ */
+export class MarkdownLanguageAdapter implements LanguageAdapter {
+  type = "markdown" as const;
+
+  lastQuotePrefix: PrefixKind = "";
+
+  transformIn(pythonCode: string): [string, number] {
+    if (!this.isSupported(pythonCode)) {
+      throw new Error("Not supported");
+    }
+
+    for (const [start, regex] of regexes) {
+      const match = pythonCode.match(regex);
+      if (match) {
+        const innerCode = match[1].trim();
+
+        const [quotePrefix, quoteType] = splitQuotePrefix(start);
+        // store the quote prefix for later when we transform out
+        this.lastQuotePrefix = quotePrefix;
+        const unescapedCode = innerCode.replaceAll(`\\${quoteType}`, quoteType);
+
+        const offset = pythonCode.indexOf(innerCode);
+        return [dedent(unescapedCode), offset];
+      }
+    }
+
+    return [pythonCode, 0];
+  }
+
+  transformOut(code: string): [string, number] {
+    // Get the quote type from the last transformIn
+    const prefix = upgradePrefixKind(this.lastQuotePrefix, code);
+
+    const isOneLine = !code.includes("\n");
+    const escapedCode = code.replaceAll('"""', '\\"""');
+    if (isOneLine) {
+      const start = `mo.md(${prefix}"""`;
+      const end = `""")`;
+      return [start + escapedCode + end, start.length];
+    }
+
+    // Multiline code
+    const start = `mo.md(\n\t${prefix}"""\n`;
+    const end = `\n\t"""\n)`;
+    return [start + indentOneTab(escapedCode) + end, start.length + 1];
+  }
+
+  isSupported(pythonCode: string): boolean {
+    return regexes.some(([, regex]) => regex.test(pythonCode));
```

Comment:
Actually the check was helpful. Without the check now we match cells containing two mo.md() lines, which then garbles the code when you switch to markdown and back.

How about this?

```suggestion
    const markdownLines = pythonCode                                             
        .split("\n")                                                               
        .map((line) => line.startsWith("mo.md("))                                  
        .filter(Boolean);                                                          
    if (markdownLines.length > 1) {                                              
        // more than line starting with mo.md(; as a heuristic,                    
        // don't show "view as markdown"                                           
        return false;                                                              
    }                 
    return regexes.some(([, regex]) => regex.test(pythonCode));
```